### ISSUE TITLE
[VL] Spark function support list [please leave comment/mark if you plan to implement]

### ISSUE THREAD

**Issue by PHILO-HE** [oap-project/gluten#4039]
### Description

Here listed some functions we need to support in Gluten and Velox, will keep updated in the future. You may reply in the thread if you'd like to work on anyone.

- [x] size https://github.com/facebookincubator/velox/pull/5461 fix return type
- [ ] percentile_approx
- [ ] explode
- [ ] concat_ws
- [x] collect_list (velox supported, needs Gluten to enable array for project plan node)
- [ ] date_sub (one velox upstream PR near to merge)
- [ ] date_format: can map to format_datatime https://github.com/facebookincubator/velox/issues/5734
- [x] regexp_replace (supported, only few incompatible issue from RE2 used by velox)
- [x] collect_set
- [x] current_timestamp (foldable expression, optimized by spark, no need to offload)
- [ ] unix_timestamp: "Only supports string type, with session timezone considered, todo: support date type"
- [ ] to_json
- [x] str_to_map https://github.com/facebookincubator/velox/pull/5842
- [ ] json_tuple
- [ ] to_date
- [x] rank
- [ ] hour
- [ ] locate
- [ ] mod
- [ ] conv: "PR ready“
- [ ] parse_url
- [ ] pow
- [ ] ifnull
- [ ] translate (upstream velox PR merged, Gluten PR pending for future velox rebase)
- [ ] urldecoder: "UDF, supported by spark as a built-in funciton since 3.4.0."
- [ ] distinct
- [ ] normalizenanandzero
- [ ] isarrayintersects
- [ ] default.json_split (udf, no need to impl.): "external UDF"
- [ ] parsejsonarray: "external UDF"
- [x] <=> (EqualNullSafe) https://github.com/oap-project/velox/pull/357 https://github.com/oap-project/gluten/pull/2259
- [ ] struct
- [ ] rand (supported except seed is provided)
- [x] now (foldable expression, optimized by spark, no need to offload)
- [ ] lead: "PR ready"
- [ ] add_months
- [ ] percentile
- [ ] last_day
- [ ] lag: "PR ready"
- [ ] minute
- [ ] year
- [ ] month
- [ ] weekofyear
- [ ] next_day
- [ ] posexplode
- [ ] trunc
- [ ] months_between
- [ ] array_min
- [ ] dense_rank
- [x] current_date (foldable expression, optimized by spark, no need to offload)
- [ ] day
- [ ] approx_percentile: "Third argument accuracy is different with velox, velox is double but spark is long"
- [ ] reflect
- [ ] date_trunc
- [ ] dayofweek
- [ ] stack
- [ ] nullif
- [ ] grouping_id
- [ ] printf
- [ ] space
- [x] bin
- [ ] find_in_set
- [ ] inline
- [ ] hex
- [ ] dayofmonth
- [ ] to_unix_timestamp: "Only supports string type, with session timezone considered. todo: support date type"

**Comment by Yohahaha** [comment]
I'd like support hex and uhex.

**Comment by zwangsheng** [comment]
Hi i'd like to give a try with hour function.

**Comment by konjac** [comment]
Hi, I'd like to have a look into `map_keys`

**Comment by fyp711** [comment]
Hi I'd like to support `find_in_set`  in velox

**Comment by HannanKan** [comment]
Hi, I'd like to support date_trunc/trunc.

**Comment by JkSelf** [comment]
Hi, I'd like to support `dense_rank`.

**Comment by JkSelf** [comment]
`dense_rank ` already supported in velox https://github.com/facebookincubator/velox/pull/6289.

**Comment by zhztheplayer** [comment]
>  - [ ] percentile_approx

> - [ ] approx_percentile: "Third argument accuracy is different with velox, velox is double but spark is long"

The two stand for the same function I assume? I'll take these two if nobody is working on it.

**Comment by PHILO-HE** [comment]
> > * [ ]  percentile_approx
> 
> > * [ ]  approx_percentile: "Third argument accuracy is different with velox, velox is double but spark is long"
> 
> The two stand for the same function I assume? I'll take these two if nobody is working on it.

Yes, they are one thing. Just unify them into one checkbox. Thanks!

**Comment by JkSelf** [comment]
I will take a look `ntile ` window function.

**Comment by zhouyuan** [comment]
ubase64: 
https://github.com/oap-project/gluten/pull/4482

**Comment by zjuwangg** [comment]
Is there any plan to suppport from_json function?

**Comment by yma11** [comment]
I'd like take `map_entries` and `map_from_entries`, there are already presto implementation in velox, will need check consistency .

**Comment by PHILO-HE** [comment]
Just removed the below functions from the list, since they have been supported. Thanks! @acvictor, @Yohahaha, @fyp711, @zwangsheng, @JkSelf, etc.
```
to_date hour mod pow ifnull add_months next_day dense_rank find_in_set hex ntile
date_from_unix_date array_repeat array_position array_except array_distinct weekday
```

**Comment by acvictor** [comment]
@PHILO-HE I also see support for `year`, `month`, `day`, `last_day` in Velox too. I can also give `from_utc_timestamp` a go.

**Comment by Surbhi-Vijay** [comment]
`nullif` is out of the box supported. Spark send the converted expression as `If` expression and it is supported in Gluten.

**Comment by PHILO-HE** [comment]
> `nullif` is out of the box supported. Spark send the converted expression as `If` expression and it is supported in Gluten.

Thanks so much for your feedback! Just removed it from the list.

**Comment by acvictor** [comment]
> @PHILO-HE I see support for `year`, `month`, `day`, `last_day` in Velox too. I can also give `from_utc_timestamp` a go.

Will do minute as well.

**Comment by rui-mo** [comment]
I'd like to work on locate and normalizenanandzero.

**Comment by mskapilks** [comment]
I would like to work on `bool_and`, `bool_or`

**Comment by JkSelf** [comment]
I'd like to work on `first_value `and `last_value `window function.

**Comment by zhztheplayer** [comment]
>  - [x] collect_list (velox supported, needs Gluten to enable array for project plan node)
>  - [x] collect_set

@PHILO-HE Should we uncheck these two? I ran a test and the two functions are both fallen back.

**Comment by Surbhi-Vijay** [comment]
I would like to give `printf` a try.

**Comment by mskapilks** [comment]
> I would like to work on `bool_and`, `bool_or`

These are already supported it seems. All `bool_and`, `bool_or`, `every`, `some` get converted to min, max of bool column

**Comment by yma11** [comment]
@PHILO-HE I would like to take `map_filter`.  Thanks.

**Comment by supermem613** [comment]
I'd like to pick up base64 and unbase64, please. 

(FYI, looks like there was a PR above for unbase64, but it seems to have been closed without committing ~45-55 days ago, so hopefully I am not conflicting with any work).

**Comment by PHILO-HE** [comment]
> @PHILO-HE , I'd like to pick up base64 and unbase64, please.
> 
> (FYI, looks like there was a PR above for unbase64, but it seems to have been closed without committing ~45-55 days ago, so hopefully I am not conflicting with any work).

Hi @supermem613, sorry for the late reply. I note Gluten PR https://github.com/apache/incubator-gluten/pull/5242 is trying to re-use Velox's existing `from_base64` function (proposed for prestosql) for `unbase64`. Not sure whether we can map `base64` to some other function. If there is no semantic difference, we can just re-use the existing Velox functions.

**Comment by PHILO-HE** [comment]
Just removed the below supported functions from the above list. Thanks for the contribution!
`last_day, unhex, lead, lag, minute, second, may_keys`

**Comment by Yohahaha** [comment]
I'd like take percentile agg function.

**Comment by supermem613** [comment]
Hey, @PHILO-HE , what's the plan with concat_ws ? It says "PR Ready" and I see that you have committed an implementation to this branch: https://github.com/oap-project/velox/commit/c5eec030464970b83389c598354c0da4c8fb25ef. Is that the PR being referenced? Is it planned to be merged into velox main?

**Comment by PHILO-HE** [comment]
> Hey, @PHILO-HE , what's the plan with concat_ws ? It says "PR Ready" and I see that you have committed an implementation to this branch: [oap-project/velox@c5eec03](https://github.com/oap-project/velox/commit/c5eec030464970b83389c598354c0da4c8fb25ef). Is that the PR being referenced? Is it planned to be merged into velox main?

Hi @supermem613, that commit is not used by Gluten main branch. We have another implementation for upstream velox: https://github.com/facebookincubator/velox/pull/8854. It is still under review. I will try to push the progress.

**Comment by Yohahaha** [comment]
I'd like take get function, as known as GetArrayItem.

**Comment by Yohahaha** [comment]
I'd like take transform function.

**Comment by lyy-pineapple** [comment]
@PHILO-HE hello,  I'd like take forall function.

**Comment by acvictor** [comment]
I'd like to try array_size.

**Comment by gaoyangxiaozhu** [comment]
hey @zhouyuan  could you help add `format_string` and `format_number` in the list and I would take `format_string` and `format_number` later

**Comment by PHILO-HE** [comment]
> hey @zhouyuan could you help add `format_string` and `format_number` in the list and I would take `format_string` and `format_number` later

@gaoyangxiaozhu, just added them into the list. Thanks!

**Comment by zhli1142015** [comment]
I'd like to take `soundex` and `levenshtein`, thanks.

**Comment by honeyhexin** [comment]
I'd like to take cot, thanks.

**Comment by Donvi** [comment]
I'd like and am working in the math function expm1.

**Comment by gaoyangxiaozhu** [comment]
PR for width_bucket support, https://github.com/apache/incubator-gluten/pull/5634  looks still need velox side change for bucket_number <=0 case will send PR in velox repository to fix

**Comment by ivoson** [comment]
I'd like to implement `array_append` and `array_insert` for spark 3.4+

**Comment by marin-ma** [comment]
> I'd like to take a look at `stack` function, it seems like a `Generator`, meaning one row of input might return multiple rows of output, does Velox has this generator ability?

@xumingming Currently, 4 generator functions are supported : `explode`, `pos_explode`, 'inline` and `json_tuple`. The approach is creating a `ProjectNode` + `UnnestNode` + `ProjectNode` pattern in Velox pipeline. But seems like the `stack` function cannot use this pattern. Perhaps we can build another pipeline by leveraging the `ExpandNode` in Velox (Not sure if this approach really works).

**Comment by NEUpanning** [comment]
I'd like to take unix_date, thanks.

**Comment by PHILO-HE** [comment]
> I'd like to take unix_date, thanks.

We have supported it in both Gluten & Velox. Just changed its state in the list. Thanks!
https://github.com/apache/incubator-gluten/pull/5287
https://github.com/facebookincubator/velox/pull/8725

**Comment by NEUpanning** [comment]
@PHILO-HE Thanks for your feedback. So i'd like to take date_part. Is to_date supported in gluten now? It doesn't shows in the list. I also would like to pick it.

**Comment by PHILO-HE** [comment]
> @PHILO-HE Thanks for your feedback. So i'd like to take date_part. Is to_date supported in gluten now? It doesn't shows in the list. I also would like to pick it.

@NEUpanning, this list only maintains working-in-progress functions. I think `to_date` has been supported. See https://github.com/apache/incubator-gluten/blob/main/docs/velox-backend-support-progress.md.

`date_part` may be supported also. I note the below test in Gluten. You can confirm whether all date patterns have been supported.
https://github.com/apache/incubator-gluten/blob/d74fc97cf941759c79f440b0df5c5071655b984e/backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala#L808

**Comment by NEUpanning** [comment]
> > @PHILO-HE Thanks for your feedback. So i'd like to take date_part. Is to_date supported in gluten now? It doesn't shows in the list. I also would like to pick it.
> 
> @NEUpanning, this list only maintains working-in-progress functions. I think `to_date` has been supported. See https://github.com/apache/incubator-gluten/blob/main/docs/velox-backend-support-progress.md.
> 
> `date_part` may be supported also. I note the below test in Gluten. You can confirm whether all date patterns have been supported.
> 
> https://github.com/apache/incubator-gluten/blob/d74fc97cf941759c79f440b0df5c5071655b984e/backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala#L808

I can't find any implementation of date_part and to_date function in Velox. Would you like to help me find it? Thanks.

**Comment by PHILO-HE** [comment]
> > > @PHILO-HE Thanks for your feedback. So i'd like to take date_part. Is to_date supported in gluten now? It doesn't shows in the list. I also would like to pick it.
> > 
> > 
> > @NEUpanning, this list only maintains working-in-progress functions. I think `to_date` has been supported. See https://github.com/apache/incubator-gluten/blob/main/docs/velox-backend-support-progress.md.
> > `date_part` may be supported also. I note the below test in Gluten. You can confirm whether all date patterns have been supported.
> > https://github.com/apache/incubator-gluten/blob/d74fc97cf941759c79f440b0df5c5071655b984e/backends-velox/src/test/scala/org/apache/gluten/execution/ScalarFunctionsValidateSuite.scala#L808
> 
> I can't find any implementation of date_part and to_date function in Velox. Would you like to help me find it? Thanks.

@NEUpanning, not a direct replacement. date_part is covered [here](https://github.com/apache/incubator-gluten/blob/main/gluten-core/src/main/scala/org/apache/gluten/expression/ExpressionConverter.scala#L466).  to_date is converted to Cast + GetTimestamp by Spark.

**Comment by PHILO-HE** [comment]
> `shuffle`, `array_sort` are already supported, can be marked as complete.

Seems `sort_array` is supported, but `array_sort` is not. Please spare some time to confirm. Thanks!

**Comment by Donvi** [comment]
As I see only `rand` exists and no `randn`, I'm taking randn

**Comment by boneanxs** [comment]
@PHILO-HE I can try to support `array_sort` if no one picked, we internally need this function :)

**Comment by Donvi** [comment]
> ubase64: #4482

I see you've map the from_base64 to unbase64, and respectively I find the base64 is almost the same as to_base64, so it's just a missing or is there any other consideration?

**Comment by PHILO-HE** [comment]
> > ubase64: #4482
> 
> I see you've map the from_base64 to unbase64, and respectively I find the base64 is almost the same as to_base64, so it's just a missing or is there any other consideration?

@Donvi, seems there are a few semantic differences between Spark's `unbase64` & Velox's `from_base64`. So the simple mapping has not been accepted by the community. See discussion: https://github.com/apache/incubator-gluten/pull/5242#discussion_r1548887962. I guess similarly `to_base64` cannot be mapped due to some unknown differences.

**Comment by gaoyangxiaozhu** [comment]
FYI, i am working for mask function support.

**Comment by zhli1142015** [comment]
I'd like to pick up `mode`, thanks

**Comment by jinchengchenghh** [comment]
Can you add empty2null to the list? @PHILO-HE

**Comment by PHILO-HE** [comment]
> Can you add empty2null to the list? @PHILO-HE

Just added.

**Comment by jinchengchenghh** [comment]
Thanks!

**Comment by jinchengchenghh** [comment]
Can you add the function `toprettystring` to the list?
This query will use it
```
select        sum(hash(floor(l_extendedprice)) *l_discount + hash(l_orderkey) + hash(l_partkey) + hash(l_suppkey) + hash(l_linenumber) + hash(l_comment) + hash(l_shipinstruct)) as revenue from      lineitem;
```

**Comment by zhli1142015** [comment]
I would lie to take AtLeastNNonNulls, thanks.

**Comment by jinchengchenghh** [comment]
Here list some other functions that not support:
https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc#L62
Here list some function some data type or some behavior does not aligns with Spark.
https://github.com/apache/incubator-gluten/blob/main/cpp/velox/substrait/SubstraitToVeloxPlanValidator.cc#L188

**Comment by zml1206** [comment]
> Hi, I'd like to support date_trunc/trunc.

@HannanKan Are you still doing this? If you don't have time, I can take over, thank you.

**Comment by zjuwangg** [comment]
> @PHILO-HE I can try to support `array_sort` if no one picked, we internally need this function :)
@ boneanxs How about this issue goes? If you don't have time, I'd like to investigate in it.

**Comment by boneanxs** [comment]
> > @PHILO-HE I can try to support `array_sort` if no one picked, we internally need this function :)
> 
> @boneanxs How about this issue goes? If you don't have time, I'd like to investigate in it.

@zjuwangg can see this pr: https://github.com/facebookincubator/velox/pull/10138, still under reviewing

**Comment by wecharyu** [comment]
@PHILO-HE I'd like to support `from_json` and `to_json`.

**Comment by rui-mo** [comment]
> I'd like to support `from_json` and `to_json`.

@wecharyu Thanks for the update. We've got one PR for `from_json`: https://github.com/facebookincubator/velox/pull/11709.

**Comment by zjuwangg** [comment]
> I'd like take `map_entries` and `map_from_entries`, there are already presto implementation in velox, will need check consistency .
@yma11 
Do you have further update on the consistency issue about `map_from_entries`?

**Comment by ayushi-agarwal** [comment]
Can you add the function get_struct_field to the list? Thanks! @PHILO-HE
Even after adding support for from_json this query fallsback to Spark due to get_struct_field
val schema = new StructType().add("platformId", StringType).add("userId", StringType).add("sessionId", StringType)
    val filteredDF = parquetDF.withColumn("parsed_json", from_json(col("json_column"), schema)).filter(col("parsed_json.platformId") === "IPHONE")

**Comment by PHILO-HE** [comment]
@ayushi-agarwal, GetStructField is an internal function. I just added it in the list.

**Comment by rui-mo** [comment]
@ayushi-agarwal I assume GetStructField is supported via SelectionNode in https://github.com/apache/incubator-gluten/blob/main/backends-velox/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala#L54. Would you provide more details on the fallback reason? Thanks.

**Comment by PHILO-HE** [comment]
> > I'd like take `map_entries` and `map_from_entries`, there are already presto implementation in velox, will need check consistency .
> 
> @yma11 Do you have further update on the consistency issue about `map_from_entries`?

@zjuwangg, Ma Yan has no bandwidth on this now. I just created one Velox pr based on her work. See https://github.com/facebookincubator/velox/pull/11934.

**Comment by ayushi-agarwal** [comment]
> @ayushi-agarwal I assume GetStructField is supported via SelectionNode in https://github.com/apache/incubator-gluten/blob/main/backends-velox/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala#L54. Would you provide more details on the fallback reason? Perhaps a new issue could be opened for the unexpected fallback. Thanks.

It fails here https://github.com/apache/incubator-gluten/blob/eeca5729b8612675a88e17ba7ba1f82b1cbd3955/backends-velox/src/main/scala/org/apache/gluten/expression/ExpressionTransformer.scala#L72C7-L74C73, this is the case where child node is of ScalarFunctionNode type.
This is the expression:
24/12/23 10:00:23 WARN GlutenFallbackReporter: Validation failed for plan: Filter[QueryId=4], due to: Unsupported child expression of GetStructField: from_json(StructField(platformId,StringType,true), json_column#7, Some(America/Los_Angeles)).platformId

**Comment by zhouyuan** [comment]
> > I'd like to support `from_json` and `to_json`.
> 
> @wecharyu Thanks for the update. We've got one PR for `from_json`: [facebookincubator/velox#11709](https://github.com/facebookincubator/velox/pull/11709). Just assigned `to_json` to you.

CC @zhli1142015 for his awareness

**Comment by ayushi-agarwal** [comment]
@wecharyu Do you have a patch for to_json support? We also wanted this functionality so wanted to check with you before starting.

**Comment by boneanxs** [comment]
> @wecharyu Do you have a patch for to_json support? We also wanted this functionality so wanted to check with you before starting.

can check it here: https://github.com/facebookincubator/velox/pull/11995

**Comment by ayushi-agarwal** [comment]
I see array_join function mentioned in here https://github.com/apache/incubator-gluten/blob/main/docs/velox-backend-support-progress.md#function-support but I don't find it here in the list? Shall we add it here also @PHILO-HE ?

**Comment by PHILO-HE** [comment]
> I see array_join function mentioned in here https://github.com/apache/incubator-gluten/blob/main/docs/velox-backend-support-progress.md#function-support but I don't find it here in the list? Shall we add it here also [@PHILO-HE](https://github.com/PHILO-HE) ?

@ayushi-agarwal, this list is for tracking unsupported or working in progress functions, not for listing all Spark functions and their status.

**Comment by ayushi-agarwal** [comment]
@Yohahaha Are you working on percentile function support?

**Comment by zjuwangg** [comment]
We may also  can support Array_size function  https://spark.apache.org/docs/latest/api/sql/index.html#array_size

Since velox has supported this spark function https://github.com/facebookincubator/velox/pull/9516

I'd like to take the support work!

**Comment by mudit-97** [comment]
@PHILO-HE I would like to support from_csv in gluten, hope thats fine

**Comment by wecharyu** [comment]
@PHILO-HE `map_concat` is ready in velox, I'd like to implement it in gluten. cc: @rui-mo

**Comment by wecharyu** [comment]
> `map_concat` is ready in velox, I'd like to implement it in gluten

Sorry, `map_concat` is already available in https://github.com/apache/incubator-gluten/pull/8781.

**Comment by minni31** [comment]
I would be picking up grouping_id, schema_of_csv, schema_of_json, to_csv, make_interval, make_dt_interval
